drbg/src/hashdrbg.rs

```diff
@@ -0,0 +1,220 @@
+    /// during instantiation or reseeding.
+    pub(crate) reseed_counter: u64,
+}
+
+impl<Hash: Digest + HashDrbgProps> HashDrbg<Hash> {
+    /// Generic hash function defined by `Hash` type
+    fn hash(data: impl Iterator<Item = impl AsRef<[u8]>>) -> Output<Hash> {
+        let mut hasher = Hash::new();
+        for item in data {
+            hasher.update(item);
+        }
+        hasher.finalize()
+    }
+    /// This is `Hash_df()` from _10.3.1_
+    ///
+    /// This function omits the number of bits to return, it always returns an array of
+    /// the length [HashDrbgProps::SeedLenBytes]
+    fn hash_df(data: &[&[u8]]) -> GenericArray<u8, Hash::SeedLenBytes> {
+        let mut tmp = GenericArray::default();
+        let mut counter: u8 = 1;
+        let len = Hash::SeedLenBytes::USIZE / Hash::OutputSize::USIZE;
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
+        for i in 0..len {
+            let counter_byte: &[u8] = &[counter];
+            let hashables = once(&counter_byte)
+                .chain(once(&no_of_bits_to_return))
+                .chain(data.iter());
+            let out = Self::hash(hashables);
+            tmp[i * Hash::OutputSize::USIZE..(i + 1) * Hash::OutputSize::USIZE]
+                .copy_from_slice(&out);
+            counter += 1;
+        }
+        let remainder = Hash::SeedLenBytes::USIZE % Hash::OutputSize::USIZE;
+        // we may need to do the hashing on last time
+        if remainder != 0 {
+            let counter_byte: &[u8] = &[counter];
+            let hashables = once(&counter_byte)
+                .chain(once(&no_of_bits_to_return))
+                .chain(data.iter());
+            let out = Self::hash(hashables);
+            tmp[len * Hash::OutputSize::USIZE..].copy_from_slice(&out[..remainder]);
+        }
+        tmp
+    }
+    /// This is the `Hash_DRBG_Instantiate_algorithm()` from _10.1.1.2_.
+    ///
+    /// Note that security strength is not used in the
+    /// implementation so it is not passed here.
+    fn instantiate_inner(
+        entropy_input: &[u8],
+        nonce: &[u8],
+        personalization_string: &[u8],
+    ) -> Self {
+        let v = Self::hash_df(&[entropy_input, nonce, personalization_string]);
+        let c = Self::hash_df(&[&[0], v.as_slice()]);
+        Self {
+            v,
+            c,
+            reseed_counter: 1,
+        }
+    }
+    /// This is the `Hash_DRBG_Reseed_algorithm()` from _10.1.1.3_.
+    fn reseed_inner(&mut self, entropy_input: &[u8], additional_input: &[u8]) {
+        self.v = Self::hash_df(&[&[0x1], self.v.as_slice(), entropy_input, additional_input]);
+        self.c = Self::hash_df(&[&[0x0], self.v.as_slice()]);
+        self.reseed_counter = 1;
+    }
+    /// This is the `Hash_DRBG_Generate_algorithm()` from _10.1.1.4_.
+    fn generate_inner(&mut self, additional_input: &[u8], out: &mut [u8]) {
+        let p = U896::from_u64(2).shl(Hash::SeedLenBytes::USIZE * 8);
```

Comment:
We want to create the BigNum U896 having value 2^x to use for modulo operation.
We can write that:  (2<<x) = 2^(x+1) or (1<<x) = 2^x.
So more precisely we may want to do:
p = U896::from_u64(1).shl(Hash::SeedLenBytes::USIZE * 8);

Reply:
Or if allowed: U896::ONE.shl(Hash::SeedLenBytes::USIZE * 8);
(I can't deduce if the ONE constant must be copied first before shl(...), or if shl(...) will return a new instance)

Reply:
nice catch! I have no idea how I missed that and how did that pass tests!